psutil_agent/loop.py

```diff
@@ -0,0 +1,115 @@
+
+    credentials = pika.PlainCredentials(rabbit_config['broker_user'],
+                                        rabbit_config['broker_pass'])
+
+    params = pika.ConnectionParameters(credentials=credentials,
+                                       host=rabbit_config['broker_hostname'],
+                                       virtual_host=rabbit_config['vhost'],
+                                       port=rabbit_config['broker_port'])
+
+    connection = pika.BlockingConnection(params)
+    channel = connection.channel()
+    channel.exchange_declare(exchange='cloudify-monitoring',
+                             exchange_type="topic",
+                             auto_delete=True,
+                             durable=False,
+                             internal=False)
+
+    for config in psutil_config:
+        def create_scheduled_fun(method, interval, f_args, result_argument,
+                                 alias):
+            try:
+                fun = getattr(psutil, method)
+
+                def scheduled_fun():
+                    try:
+                        result = fun(**f_args)
+                    except Exception as e:
+                        result = None
+                        logging.error(
+                                'Invoking a psutil function failed: {0} {1}'
+                                .format(type(e).__name__, e))
+
+                    if result:
+                        if result_argument:
+                            result = getattr(result, result_argument)
+
+                        service_elements = [
+                            rabbit_config['deployment_id'],
+                            rabbit_config['node_name'],
+                            rabbit_config['node_id'],
+                            alias or method
+                        ]
+
+                        metric_data = {
+                            'node_id': rabbit_config['node_id'],
+                            'node_name': rabbit_config['node_name'],
+                            'deployment_id': rabbit_config['deployment_id'],
+                            'name': alias or method,
+                            'path': '',
+                            'metric': result,
+                            'unit': '',
+                            'type': 'GAUGE',
+                            'host': rabbit_config['node_id'],
+                            'service': '.'.join(service_elements),
+                            'time': int(time.time()),
+                        }
+
+                        channel.basic_publish(
+                                exchange='cloudify-monitoring',
+                                routing_key=rabbit_config['deployment_id'],
+                                body=json.dumps(metric_data))
+
+                    scheduler.enter(interval, 1, scheduled_fun, ())
+
+                scheduled_fun()
+
+            except AttributeError as e:
+                logging.error(e)
+            except AMQPError as e:
```

Comment:
i think this should be only on the "basic_publish" method, and if it fails you want to reenter the loop